### PR TITLE
url: don't set appconnect time for non-ssl/non-ssh connections

### PR DESCRIPTION
**Pull request by jay**
Prior to this change non-ssl/non-ssh connections that were reused set
TIMER_APPCONNECT [1]. Arguably that was incorrect since no SSL/SSH
handshake took place.

[1]: TIMER_APPCONNECT is publicly known as CURLINFO_APPCONNECT_TIME in
libcurl and %{time_appconnect} in the curl tool. It is documented as
"the time until the SSL/SSH handshake is completed".

Reported-by: Marcel Hernandez

Ref: https://github.com/curl/curl/issues/3760

Closes #xxxx

---

I'm not sure about the SSH check. At that point is it certain that a handler exists, and if it's SSH family that authed will be true if the handshake is complete?

Further, refer to #3760 for different views. As @1ma [points out](https://github.com/curl/curl/issues/3760#issuecomment-482315653) we have a chart that shows appconnect time coming after connect. Yet appconnect is [documented](https://curl.haxx.se/libcurl/c/CURLINFO_APPCONNECT_TIME.html) to "get the time until the SSL/SSH handshake is completed".

### CODE REVIEW

**Comment by 1ma** [comment]
I'd rather see this fixed the other way around, for the reasons exposed in the issue. But if it's not, I _think_ that then reused HTTPS connections should also report 0 appconnect time, since no new SSL handshake seems to take place in that case, too:

```
$ curl -s -w@curl-format.txt -o /dev/null https://www.google.com/ -o /dev/null https://www.google.com/
    time_namelookup:  0.001085
       time_connect:  0.010285
    time_appconnect:  0.063561
   time_pretransfer:  0.063676
 time_starttransfer:  0.126158
                    ----------
         time_total:  0.128053

    time_namelookup:  0.000031
       time_connect:  0.000032
    time_appconnect:  0.000032
   time_pretransfer:  0.000085
 time_starttransfer:  0.080010
                    ----------
         time_total:  0.082248
```

File: lib/url.c [hunk 1]
```diff
@@ -3992,7 +3992,10 @@ CURLcode Curl_setup_conn(struct connectdata *conn,
   }
   else {
     Curl_pgrsTime(data, TIMER_CONNECT);    /* we're connected already */
-    Curl_pgrsTime(data, TIMER_APPCONNECT); /* we're connected already */
+    if(conn->ssl[FIRSTSOCKET].use ||
+       ((conn->handler->protocol & PROTO_FAMILY_SSH) &&
+        conn->proto.sshc.authed))
```
Comment (bagder) [inline lib/url.c]: I don't think "conn->proto.sshc.authed" is necessary here. If the SSH protocol family is used, there *will* be an appconnect phase.

Comment (jay) [reply lib/url.c]: > I don't think "conn->proto.sshc.authed" is necessary here. If the SSH protocol family is used, there _will_ be an appconnect phase.

In the code appconnect timer is only set when authed is complete. So there will be a phase but could the connection be reused without authed being complete?

**Comment by jay** [comment]
> . But if it's not, I _think_ that then reused HTTPS connections should also report 0 appconnect time, since no new SSL handshake seems to take place in that case, too:

Sure I understand what you're saying. Technically time-from-start is almost instant on a reused connection. So connect and namelookup and appconnect on a reused connection are all going to be a very short time like 0.000032. Whether it's correct to report that or 0 I think is debatable. This commit doesn't address that it just addresses setting appconnect when there is no handshake, which in my opinion is not correct.

File: lib/url.c (hunk 1 above)
Comment (bagder) [reply lib/url.c]: No, it shouldn't be possible. When the connection is put into the connection pool, it is either authenticated or it failed (and thus wasn't put into the pool to begin with).

**Comment by bagder** [comment]
Merge before release?

**Comment by jay** [comment]
> Merge before release?

Oops didn't see this until just now. Still for it but I'd like to get a +1.

**Pull request closed without merging**